hal-core/src/addr.rs

```diff
@@ -13,16 +13,40 @@ pub trait Address:
     + fmt::Debug
 {
     fn as_usize(self) -> usize;
+    fn from_usize(u: usize) -> Self;
 
     /// Aligns `self` up to `align`.
     ///
     /// The specified alignment must be a power of two.
-    fn align_up<A: Into<usize>>(self, align: A) -> Self;
+    ///
+    /// # Panics
+    ///
+    /// - If `align` is not a power of two.
+    fn align_up<A: Into<usize>>(self, align: A) -> Self {
+        let align = align.into();
+        assert!(align.is_power_of_two());
+        let mask = align - 1;
+        let u = self.as_usize();
+        if u & mask == 0 {
+            return self;
+        }
+        let aligned = (u | mask) + 1;
+        Self::from_usize(aligned as usize)
+    }
 
     /// Aligns `self` down to `align`.
     ///
     /// The specified alignment must be a power of two.
-    fn align_down<A: Into<usize>>(self, align: A) -> Self;
+    ///
+    /// # Panics
+    ///
+    /// - If `align` is not a power of two.
+    fn align_down<A: Into<usize>>(self, align: A) -> Self {
```

Comment:
We can totally do this in a different patch, but it might be nice to have `align_up_for` and `align_down_for` helper methods like the `is_aligned_for` method.